### ISSUE TITLE
Change Request: Support additional rule metadata for deprecations

### ISSUE THREAD

**Issue by bmish** [eslint/eslint#18061]
### ESLint version

8.56.0

### What problem do you want to solve?

There are long-time rule [properties](https://eslint.org/docs/latest/extend/custom-rules#rule-structure) `meta.deprecated` and `meta.replacedBy` that have been intended to document when rules are deprecated and what their replacement rule(s) are. For the most part, usage would look something like this:

```js
module.exports = { meta: { deprecated: true, replacedBy: ['replacement-rule-name'] } };
```

These properties are often used for generating plugin/rule documentation websites and in documentation tooling like [eslint-doc-generator](https://github.com/bmish/eslint-doc-generator).

But there are some limitations to this current format.
* Simply providing the replacement rule name as a string doesn't yield much context/explanation of the replacement/deprecation.
* Some rules provide the replacement rule name with the plugin prefix as in `prefix/rule-name` while others just provide it as `rule-name`, which can result in ambiguity about whether the replacement rule is in the same plugin, a different third-party plugin, or ESLint core. For third-party plugins, there's no easy way to automatically determine where their documentation is located or how to link to them.

### What do you think is the correct solution?

I would like to propose an extended `meta.deprecated` / `meta.replacedBy` rule property schema to reduce ambiguity and allow additional key details to be represented in it, described below as a TypeScript type for clarity:

```ts
type RuleMeta = {
  deprecated:
    | boolean // Existing boolean option.
    | string // General deprecation message, such as why the deprecation occurred.
    | {
        message?: string; General deprecation message, such as why the deprecation occurred.
        url?: string; // URL to more information about this deprecation.
      };
  replacedBy:
    | readonly string[] // Existing string array option of rule names.
    | readonly {
        /**
         * Plugin containing the replacement.
         * Use "eslint" if the replacement is an ESLint core rule.
         * Omit if the replacement is in the same plugin.
         */
        plugin?:
          | string // Plugin name i.e. "eslint-plugin-example" that contains the replacement rule.
          | {
              name?: string; // Plugin name i.e. "eslint-plugin-example" that contains the replacement rule.
              url?: string; // URL to plugin documentation.
            };
        rule:
          | string // Replacement rule name (without plugin prefix).
          | {
              name?: string; // Replacement rule name (without plugin prefix).
              url?: string; // URL to rule documentation.
            };
        deprecation?: {
          message?: string; // Message about this specific replacement, such as, how to use/configure the replacement rule to achieve the same results as the rule being replaced
          url?: string; // URL to more information about this deprecation.
        };
      }[]
    | undefined;
};
```

Real-world example of how this could be used based on the situation in https://github.com/eslint/eslint/issues/18053:

```js
module.exports = {
  meta: {
    deprecated: {
      message: 'Stylistic rules are being moved out of ESLint core.',
      url: 'https://eslint.org/blog/2023/10/deprecating-formatting-rules/',
    },
    replacedBy: [
      {
        plugin: {
          name: '@stylistic/js',
          url: 'https://eslint.style/',
        },
        rule: {
          name: 'semi',
          url: 'https://eslint.style/rules/js/semi',
        },
      },
    ],
  },
};
```

This data could be used by documentation websites and tooling like [eslint-doc-generator](https://github.com/bmish/eslint-doc-generator) to generate notices like:

> semi (deprecated)
> Replaced by [semi](https://eslint.style/rules/js/semi) from [@stylistic/js](https://eslint.style/).
> Stylistic rules are being moved out of ESLint core. [Read more](https://eslint.org/blog/2023/10/deprecating-formatting-rules/).

We can also support the same `meta.deprecated` and `meta.replacedBy` properties on configurations and processors (the other kinds of objects exported by ESLint plugins), replacing `rule` with `config` or `processor` as needed. 

In terms of actual changes needed for this:
* Mention the new schema in the custom rule [documentation](https://eslint.org/docs/latest/extend/custom-rules#rule-structure)
* Update the [type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b77d83e019025017b06953907cb77f35e4231714/types/eslint/index.d.ts#L734) in @types/eslint
* Add any additional information to `meta.replacedBy` in core rules as desired
* Update ESLint's website generator to take into account the additional information

This proposal is inspired by:
* https://github.com/jsx-eslint/eslint-plugin-react/pull/3469#discussion_r1002316631
* https://github.com/eslint/eslint/issues/5774#issuecomment-220640368

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by Rec0iL99** [comment]
ping @eslint/eslint-team

**Comment by Rec0iL99** [comment]
@kecrily please tag @team when moving an issue across the board. I think folks did not get a notification for this.

**Comment by nzakas** [comment]
Yes, that's true, and "Evaluating" is meant to be when you, personally, are evaluating it. That means you should leave a comment indicating that's what is happening.

If instead your intent was to ask other team members for their opinion, then you should also leave a comment and move to "Feedback Needed."

In any event, please don't just move the issue into a different column without comment. :)

**Comment by nzakas** [comment]
My opinion: This is worth exploring, but we should do it through an RFC to make sure it gets the appropriate amount of attention and exploration.

Would like some other opinions.

**Comment by JoshuaKGoldberg** [comment]
Is there a strong motivation to have two properties here? As in, is there a time when a rule would reasonably have `replacedBy` but not `deprecated`? I don't think I've ever seen that in the wild.

If the `deprecated` property is getting more metadata and becoming an object, would that RFC be a good place to discuss making `replacedBy` a property of that object?

**Comment by bmish** [comment]
If we were designing this from scratch, then yeah I'd likely put everything inside the `deprecated` object. I'm leaning toward keeping the two properties for backwards-compatibility, but this potential redesign is a good call out that I could mention in the "Alternatives" section of the RFC.

**Comment by bmish** [comment]
Thinking about it some more, perhaps I can just leave `replacedBy` alone as a string array for backwards-compat, possibly deprecating it, while adding the more sophisticated object array version of it as a new `replacedBy` property inside the `deprecated` object for the purpose of consolidating things.

**Comment by JoshuaKGoldberg** [comment]
The only other point that comes to mind for me is being able to specify options in the new rule. https://github.com/typescript-eslint/tslint-to-eslint-config -> https://github.com/typescript-eslint/tslint-to-eslint-config/blob/470d44de20beb7c7366de993edb8898d0766b8aa/docs/Architecture/Linters.md might be a good reference. Deprecated rules could map to multiple new ones, including specifying options and changed or missing functionality.

Not suggesting rules should include a transformer function themselves - just posting as examples of what information users _might_ want to know.

**Comment by nzakas** [comment]
It seems like there's interest in pursuing further, so marking as accepted and moving to "Waiting for RFC".

**Comment by nzakas** [comment]
The RFC for this has been approved: https://github.com/eslint/rfcs/pull/124

@DMartens are you interested in implementing the change?

**Comment by DMartens** [comment]
Yes, I already started and only need to update the documentation.
